### PR TITLE
Fix it

### LINKED ISSUE #491: Update jersey with the fix
@marcuslinke code mentions workaround for https://java.net/jira/browse/JERSEY-2852
https://github.com/jersey/jersey/pull/164 that was already resolved in released versions https://github.com/jersey/jersey/releases/tag/2.22.2 .

@marcuslinke could you review and update it? Maybe it will help with my TTY/hijack issues...

### PR DESCRIPTION
**Pull request by KostyaSha**
trying fix ITs with mass changes
closes #491 

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/620)

### CODE REVIEW

File: src/main/java/com/github/dockerjava/jaxrs/async/AbstractCallbackNotifier.java [hunk 1]
```diff
@@ -60,8 +60,7 @@ public Void call() throws Exception {
             return null;
         }
 
-        try {
-            InputStream inputStream = new WrappedResponseInputStream(response);
+        try (InputStream inputStream = new WrappedResponseInputStream(response)) {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/async/AbstractCallbackNotifier.java]: @marcuslinke resource leak fix

File: src/test/java/com/github/dockerjava/core/command/AttachContainerCmdImplTest.java [hunk 2]
```diff
@@ -97,14 +100,24 @@ public void onNext(Frame frame) {
...
 
+        dockerClient.close();
+
         System.out.println("log: " + callback.toString());
 
         // HexDump.dump(collectFramesCallback.toString().getBytes(), 0, System.out, 0);
-
+        RuntimeException firstError = callback.getFirstError();
+        if (isEmpty(callback.toString())) {
+            throw new SkipException("com.github.dockerjava.api.exception.InternalServerErrorException: http: Hijack is incompatible with use of CloseNotifier");
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/AttachContainerCmdImplTest.java]: @marcuslinke still don't understand why hijack issues exists, it was fixed in 1.11 version by using newer go.

File: src/test/java/com/github/dockerjava/core/command/CreateContainerCmdImplTest.java [hunk 3]
```diff
@@ -191,21 +194,24 @@ public void createContainerWithVolumesFrom() throws DockerException {
...
+                .exec();
 
         LOG.info("Created container {}", container.toString());
 
         assertThat(container.getId(), not(isEmptyString()));
 
         InspectContainerResponse inspectContainerResponse = dockerClient.inspectContainerCmd(container.getId()).exec();
 
-        assertThat(Arrays.asList(inspectContainerResponse.getConfig().getEnv()), containsInAnyOrder("VARIABLE=success"));
+        assertThat(Arrays.asList(inspectContainerResponse.getConfig().getEnv()), hasItem("VARIABLE=success"));
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/CreateContainerCmdImplTest.java]: fix for unreliable test

File: src/test/java/com/github/dockerjava/core/command/UpdateContainerCmdImplTest.java [hunk 4]
```diff
@@ -84,10 +84,10 @@ public void updateContainer() throws DockerException, IOException {
 //                .withKernelMemory(52428800) Can not update kernel memory to a running container, please stop it first.
                 .exec();
 
-        // docker toolbox 1.10.1
-        assertThat(updateResponse.getWarnings(), hasSize(1));
-        assertThat(updateResponse.getWarnings().get(0),
-                is("Your kernel does not support Block I/O weight. Weight discarded."));
+        // true only on docker toolbox (1.10.1)
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/UpdateContainerCmdImplTest.java]: travis has no such error.

**Comment by KostyaSha** [comment]
@marcuslinke i tried upgrade to jersey version that contains fix for your issue, but got
```
deleteDockerContainerImage(com.github.dockerjava.netty.exec.AttachContainerCmdExecTest)  Time elapsed: 0.326 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: java.lang.IllegalStateException: Connection pool shut down
Caused by: java.lang.IllegalStateException: Connection pool shut down
```
Have you seen this error before?

**Comment by marcuslinke** [comment]
@KostyaSha No, haven't seen this before. Will try to reproduce locally...

**Comment by KostyaSha** [comment]
@marcuslinke i see a lot of different fails, trying to split everything to small portions now.

**Comment by KostyaSha** [comment]
```
18:08:28.476 INFO  c.g.d.c.AbstractDockerClientTest - ################################## STARTING attachContainerWithTTY ##################################
18:08:37.565 INFO  c.g.d.c.AbstractDockerClientTest - Created container: com.github.dockerjava.api.command.CreateContainerResponse@322124ba[id=fb84d3a2f347087279346985f9bd62551737b6a8942e1d0fc973f5264b032c23,warnings=<null>]
18:08:37.743 ERROR c.g.d.c.async.ResultCallbackTemplate - Error during callback
com.github.dockerjava.api.exception.InternalServerErrorException: http: Hijack is incompatible with use of CloseNotifier
	at com.github.dockerjava.jaxrs.filter.ResponseStatusExceptionFilter.filter(ResponseStatusExceptionFilter.java:53) ~[docker-java-3.0.0-kostyasha-3.jar:na]
	at org.glassfish.jersey.client.ClientFilteringStages$ResponseFilterStage.apply(ClientFilteringStages.java:134) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.client.ClientFilteringStages$ResponseFilterStage.apply(ClientFilteringStages.java:123) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:171) ~[jersey-common-2.11.jar:na]
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:251) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) ~[jersey-common-2.11.jar:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) ~[jersey-common-2.11.jar:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:228) ~[jersey-common-2.11.jar:na]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:424) ~[jersey-common-2.11.jar:na]
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:679) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:435) ~[jersey-client-2.11.jar:na]
	at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:338) ~[jersey-client-2.11.jar:na]
	at com.github.dockerjava.jaxrs.async.POSTCallbackNotifier.response(POSTCallbackNotifier.java:29) ~[docker-java-3.0.0-kostyasha-3.jar:na]
	at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:50) [docker-java-3.0.0-kostyasha-3.jar:na]
	at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:24) [docker-java-3.0.0-kostyasha-3.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_80]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_80]
log: 
18:08:37.760 INFO  c.g.d.c.AbstractDockerClientTest - Cleaning up temporary container fb84d3a2f347087279346985f9bd62551737b6a8942e1d0fc973f5264b032c23
18:08:37.926 INFO  c.g.d.c.AbstractDockerClientTest - Cleaning up temporary image with 1632346790
18:08:37.935 INFO  c.g.d.c.AbstractDockerClientTest - ################################## END OF attachContainerWithTTY ##################################
```

**Comment by KostyaSha** [comment]
1.10.3 on docker.sock:
```
Tests run: 301, Failures: 8, Errors: 0, Skipped: 0, Time elapsed: 520.8 sec <<< FAILURE! - in TestSuite
copyFileToContainer(com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest)  Time elapsed: 0.464 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: Error closing message content input stream.
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyFileToContainer(CopyArchiveToContainerCmdImplTest.java:57)
Caused by: java.io.IOException: Bad file descriptor at Socket[addr=0.0.0.0/0.0.0.0,port=0,localport=0][fd=java.io.FileDescriptor@1cc1806b; file=/run/docker.sock; connected=false; bound=false]
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyFileToContainer(CopyArchiveToContainerCmdImplTest.java:57)
Caused by: org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyFileToContainer(CopyArchiveToContainerCmdImplTest.java:57)
copyStreamToContainer(com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest)  Time elapsed: 0.577 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: Error closing message content input stream.
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyStreamToContainer(CopyArchiveToContainerCmdImplTest.java:66)
Caused by: java.io.IOException: Bad file descriptor at Socket[addr=0.0.0.0/0.0.0.0,port=0,localport=0][fd=java.io.FileDescriptor@145a438c; file=/run/docker.sock; connected=false; bound=false]
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyStreamToContainer(CopyArchiveToContainerCmdImplTest.java:66)
Caused by: org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.assertFileCopied(CopyArchiveToContainerCmdImplTest.java:83)
	at com.github.dockerjava.core.command.CopyArchiveToContainerCmdImplTest.copyStreamToContainer(CopyArchiveToContainerCmdImplTest.java:66)
testEventStreamTimeBound(com.github.dockerjava.core.command.EventsCmdImplTest)  Time elapsed: 31.887 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: Error closing message content input stream.
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamTimeBound(EventsCmdImplTest.java:68)
Caused by: java.io.IOException: Bad file descriptor at Socket[addr=0.0.0.0/0.0.0.0,port=0,localport=0][fd=java.io.FileDescriptor@5bd52802; file=/run/docker.sock; connected=false; bound=false]
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamTimeBound(EventsCmdImplTest.java:68)
Caused by: org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamTimeBound(EventsCmdImplTest.java:68)
testEventStreamingWithFilter(com.github.dockerjava.core.command.EventsCmdImplTest)  Time elapsed: 21.257 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: Error closing message content input stream.
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamingWithFilter(EventsCmdImplTest.java:121)
Caused by: java.io.IOException: Bad file descriptor at Socket[addr=0.0.0.0/0.0.0.0,port=0,localport=0][fd=java.io.FileDescriptor@13654292; file=/run/docker.sock; connected=false; bound=false]
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamingWithFilter(EventsCmdImplTest.java:121)
Caused by: org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamingWithFilter(EventsCmdImplTest.java:121)
testStatsStreaming(com.github.dockerjava.core.command.StatsCmdImplTest)  Time elapsed: 4.186 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: Error closing message content input stream.
	at com.github.dockerjava.core.command.StatsCmdImplTest.testStatsStreaming(StatsCmdImplTest.java:73)
Caused by: java.io.IOException: Bad file descriptor at Socket[addr=0.0.0.0/0.0.0.0,port=0,localport=0][fd=java.io.FileDescriptor@20f15fc3; file=/run/docker.sock; connected=false; bound=false]
	at com.github.dockerjava.core.command.StatsCmdImplTest.testStatsStreaming(StatsCmdImplTest.java:73)
Caused by: org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor
	at com.github.dockerjava.core.command.StatsCmdImplTest.testStatsStreaming(StatsCmdImplTest.java:73)
testStopContainer(com.github.dockerjava.core.command.StopContainerCmdImplTest)  Time elapsed: 2.401 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <0>
     but: was <137>
	at com.github.dockerjava.core.command.StopContainerCmdImplTest.testStopContainer(StopContainerCmdImplTest.java:73)
execStartNotAttachedStdin(com.github.dockerjava.netty.exec.ExecStartCmdExecTest)  Time elapsed: 0.217 sec  <<< FAILURE!
java.lang.AssertionError: The process was not finished. expected:<false> but was:<true>
	at com.github.dockerjava.netty.exec.ExecStartCmdExecTest.execStartNotAttachedStdin(ExecStartCmdExecTest.java:205)
testStopContainer(com.github.dockerjava.netty.exec.StopContainerCmdExecTest)  Time elapsed: 2.558 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <0>
     but: was <137>
	at com.github.dockerjava.netty.exec.StopContainerCmdExecTest.testStopContainer(StopContainerCmdExecTest.java:73)
Results :
Failed tests: 
  CopyArchiveToContainerCmdImplTest.copyFileToContainer:57->assertFileCopied:83 » Processing
  CopyArchiveToContainerCmdImplTest.copyStreamToContainer:66->assertFileCopied:83 » Processing
  EventsCmdImplTest.testEventStreamTimeBound:68 » Processing Error closing messa...
  EventsCmdImplTest.testEventStreamingWithFilter:121 » Processing Error closing ...
  StatsCmdImplTest.testStatsStreaming:73 » Processing Error closing message cont...
  StopContainerCmdImplTest.testStopContainer:73 
Expected: is <0>
     but: was <137>
  ExecStartCmdExecTest.execStartNotAttachedStdin:205->Assert.assertFalse:58->Assert.failNotEquals:489->Assert.fail:89 The process was not finished. expected:<false> but was:<true>
  StopContainerCmdExecTest.testStopContainer:73 
Expected: is <0>
     but: was <137>
Tests run: 301, Failures: 8, Errors: 0, Skipped: 0
```

**Comment by KostyaSha** [comment]
$ export DOCKER_HOST="tcp://127.0.0.1:2375"
$ export DOCKER_VERSION="1.11.2-0~trusty"
```
Tests run: 358, Failures: 2, Errors: 0, Skipped: 57, Time elapsed: 480.484 sec <<< FAILURE! - in TestSuite
(com.github.dockerjava.core.command.FrameReaderITest)  Time elapsed: 0.808 sec  <<< FAILURE!
java.lang.IllegalStateException: Connection pool shut down
deleteDockerContainerImage(com.github.dockerjava.netty.exec.AttachContainerCmdExecTest)  Time elapsed: 0.326 sec  <<< FAILURE!
javax.ws.rs.ProcessingException: java.lang.IllegalStateException: Connection pool shut down
Caused by: java.lang.IllegalStateException: Connection pool shut down
Results :
Failed tests: 
com.github.dockerjava.core.command.FrameReaderITest.(com.github.dockerjava.core.command.FrameReaderITest)
  Run 1: PASS
  Run 2: FrameReaderITest.canLogInOneThreadAndExecuteCommandsInAnother » IllegalState C...
  AttachContainerCmdExecTest.deleteDockerContainerImage » Processing java.lang.I...
Tests run: 301, Failures: 2, Errors: 0, Skipped: 1
```

**Comment by KostyaSha** [comment]
One IT now green https://travis-ci.org/docker-java/docker-java/jobs/143116581 ! perfect

**Comment by KostyaSha** [comment]
And unix may fail because i switched deps...

**Comment by KostyaSha** [comment]
@marcuslinke don't worry i will make clean PR, in process..

**Comment by KostyaSha** [comment]
Everything beside junix socket update was merged. closing.

**Pull request closed without merging**